### PR TITLE
cdvdgigaherz: Refactor code and initial Linux port

### PR DESCRIPTION
**Pull request by turtleli**
I can split the PR into 2 if necessary. 1st half is rearranging/improving/bug-fixing code so it's easier to port, 2nd half is portability and the actual Linux port.

**Windows**
- Non-SPTI raw sector reading code removed - doesn't work well anyway
- Disc info ioctl use minimised
- Use formatted TOC instead of raw TOC - provides less info, but the extra info isn't really necessary and it's much easier to port the code
- C++11 locks, threads, condition variables used instead of the Win32 equivalents
- Converted to Unicode project
- Compatibility with old config broken (well, there's only one setting anyway and it made porting slightly easier)

**Linux**
- Works - Requires an extra udev dependency. Config dialog perhaps isn't so great, but it works. (EDIT: I made some fixes after opening the PR that I didn't really test...)

### CODE REVIEW

File: plugins/CMakeLists.txt [hunk 1]
```diff
@@ -6,9 +6,9 @@ if(NOT TOP_CMAKE_WAS_SOURCED)
 endif()
 
 
-#if(EXISTS "${CMAKE_SOURCE_DIR}/plugins/cdvdGigaherz" AND cdvdGigaherz)
-#	add_subdirectory(cdvdGigaherz)
-#endif()
+if(EXISTS "${CMAKE_SOURCE_DIR}/plugins/cdvdGigaherz" AND cdvdGigaherz)
+	add_subdirectory(cdvdGigaherz/src)
```
Comment (willkuer) [inline plugins/CMakeLists.txt]: intendation?

File: plugins/cdvdGigaherz/src/CDVD.cpp [hunk 2]
```diff
@@ -153,107 +142,81 @@ DWORD CALLBACK keepAliveThread(PVOID param)
...
+        return -1;
 
     // open device file
-    src = new IOCtlSrc(csrc);
-
-    if (!src->IsOK()) {
-        printf(" * CDVD: Error opening source.\n");
+    try {
+        src = std::unique_ptr<IOCtlSrc>(new IOCtlSrc(drive));
+    } catch (std::runtime_error &ex) {
```
Comment (willkuer) [inline plugins/cdvdGigaherz/src/CDVD.cpp]: Do you typically put `catch` on the same line? Looks somehow odd to me. But probably it is somehow consistent...

File: plugins/cdvdGigaherz/src/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,56 @@
+#Check that people use the good file
+if (NOT TOP_CMAKE_WAS_SOURCED)
+    message(FATAL_ERROR "
+    You did not 'cmake' the good CMakeLists.txt file. Use the one in the top dir.
+    It is advice to delete all wrongly generated cmake stuff => CMakeFiles & CMakeCache.txt")
```
Comment (willkuer) [inline plugins/cdvdGigaherz/src/CMakeLists.txt]: advice -> advised

**Comment by willkuer** [comment]
Linux community will love you.

File: plugins/CMakeLists.txt (hunk 1 above)
Comment (turtleli) [inline plugins/CMakeLists.txt]: This file mainly uses tab indentation.

File: plugins/cdvdGigaherz/src/CDVD.cpp (hunk 2 above)
Comment (turtleli) [inline plugins/cdvdGigaherz/src/CDVD.cpp]: I prefer it this way (clang-format is set up this way too). Guess it depends what style you prefer.

File: plugins/CMakeLists.txt (hunk 1 above)
Comment (turtleli) [inline plugins/CMakeLists.txt]: Oh, I thought you were initially talking about tabs/spaces and not the random "tab indentation sometimes takes a vacation on Github" thing. I've seen that a few times myself.

**Comment by gregory38** [comment]
I'm on vacation. I will look at it when I come back (Thuesday). I'm quite eager to test it :)

**Comment by FlatOutPS2** [comment]
> I'm on vacation. I will look at it when I come back (Thuesday). I'm quite eager to test it :)

Tuesday or Thursday? :D One letter has to be wrong, hard to tell which one. :D

I really need this plugin if I'm to switch to Linux sometime next year, so I'm really glad to see it. :)

**Comment by gregory38** [comment]
I meant Tuesday

**Comment by turtleli** [comment]
Ok. No rush.

I still need to figure out what to do with the cmake udev stuff anyway (I just realised Travis CI skipped the cdvdgigaherz build on my latest push). Probably just need to write or find a cmake module somewhere.

**Comment by gregory38** [comment]
There is a macro to find package based on pkg-config (you will notice there are 2 syntax to seach lib in SearchForStuff). You can likely use it.

**Comment by gregory38** [comment]
=> check_lib

**Comment by turtleli** [comment]
> => check_lib

That's what I used. But thinking about it, it might be a problem with the eudev package on my end (I use Slackware, though I've been thinking about whether I should change my distro). Will need to investigate.

**Comment by turtleli** [comment]
Oh. Ignore previous post. Seems it was because I didn't clean build so I had random link error...

**Comment by mirh** [comment]
> Non-SPTI raw sector reading code removed - doesn't work well anyway

#648 or.. ?

**Comment by turtleli** [comment]
Non-SPTI = not the recently added raw sector reading code.

**Comment by mirh** [comment]
No, I meant: when you say "not working well" are you referring to the bug I was having?

**Comment by turtleli** [comment]
Yes? And reads over multiple tracks of different types (of course, part of that is because I couldn't be bothered modifying that code as well, but I can't say I'm particularly motivated to do so since it'll still be less compatible than the SPTI code and that would feel like a waste of effort).

**Comment by turtleli** [comment]
I think I've managed to fixed the udev issue. Or so I hope. Basically pkg-config isn't used during cross compilation. so I had to search for libudev when not cross compiling (the pkg-config file is libudev.pc, not udev.pc) and udev when cross compiling.

File: plugins/cdvdGigaherz/src/ReadThread.cpp [hunk 4]
```diff
@@ -136,10 +143,7 @@ DWORD CALLBACK cdvdThread(PVOID param)
             continue;
         }
 
-        if (prefetch_left)
-            WaitForSingleObject(hNotify, 1);
-        else
-            WaitForSingleObject(hNotify, 250);
+        s_notify_cv.wait_for(guard, std::chrono::milliseconds(prefetch_left ? 1 : 250));
```
Comment (gregory38) [inline plugins/cdvdGigaherz/src/ReadThread.cpp]: What will happen in case of spurious wake up ?

File: plugins/cdvdGigaherz/src/CDVD.cpp [hunk 5]
```diff
@@ -148,35 +148,30 @@ DWORD CALLBACK keepAliveThread(PVOID param)
...
-    SetThreadPriority(hThread_keepAlive, THREAD_PRIORITY_NORMAL);
-
-    return 0;
+    return s_keepalive_is_open;
 }
 
 void StopKeepAliveThread()
 {
+    if (!s_keepalive_thread.joinable())
+        return;
```
Comment (gregory38) [inline plugins/cdvdGigaherz/src/CDVD.cpp]: I would prefer a white line after the if block.

File: plugins/cdvdGigaherz/src/CDVD.h [hunk 6]
```diff
@@ -123,4 +123,9 @@ inline void lba_to_msf(s32 lba, u8 *m, u8 *s, u8 *f)
     *f = static_cast<u8>(lba % 75);
 }
 
+#if defined(_WIN32)
+#define EXPORT
+#else
+#define EXPORT extern "C" __attribute__((stdcall, externally_visible, visibility("default")))
```
Comment (gregory38) [inline plugins/cdvdGigaherz/src/CDVD.h]: Maybe CALLBACK and EXPORT could be fused together. It is up to you.

File: plugins/cdvdGigaherz/src/ReadThread.cpp (hunk 4 above)
Comment (turtleli) [inline plugins/cdvdGigaherz/src/ReadThread.cpp]: It should be safe? Well, sort of, I think there's actually a pre-existing data race that I haven't actually dealt with (On master, 1.4.0, and this PR I managed to get some TLB misses in Growlanser: Heritage Of War when using the plugin, but it works perfectly fine on ISO).

**Comment by gregory38** [comment]
Not sure it is related but could be a wrong API setup (stdcall and such). In doubt, I prefer to keep a trace.
```
=================================================================
==4135==ERROR: AddressSanitizer: stack-buffer-overflow on address 0xd3bfef84 at pc 0x897794c bp 0xd3bfeef8 sp 0xd3bfeeec
READ of size 4 at 0xd3bfef84 thread T7 (MTGS)
    #0 0x897794b in Threading::pxThread::_try_virtual_invoke(void (Threading::pxThread::*)()) ../common/src/Utilities/ThreadTools.cpp:590
    #1 0x8977bfc in Threading::pxThread::_ThreadCleanup() ../common/src/Utilities/ThreadTools.cpp:597
    #2 0x897c983 in Threading::pxThread::_pt_callback_cleanup(void*) ../common/src/Utilities/ThreadTools.cpp:169
    #3 0x897c983 in ~__pthread_cleanup_class /usr/include/pthread.h:572
    #4 0x897c983 in Threading::pxThread::_internal_callback(void*) ../common/src/Utilities/ThreadTools.cpp:685
    #5 0xf723e51f (/usr/lib/i386-linux-gnu/libasan.so.1+0x5651f)
    #6 0xf7206077 (/usr/lib/i386-linux-gnu/libasan.so.1+0x1e077)
    #7 0xf5abbefa in start_thread (/lib/i386-linux-gnu/i686/cmov/libpthread.so.0+0x6efa)
    #8 0xf59f3d0d in clone (/lib/i386-linux-gnu/i686/cmov/libc.so.6+0xebd0d)

Address 0xd3bfef84 is located in stack of thread T7 (MTGS) at offset 84 in frame
    #0 0x897b7cf in Threading::pxThread::_internal_execute() ../common/src/Utilities/ThreadTools.cpp:633

  This frame has 1 object(s):
    [32, 44) 'lock' <== Memory access at offset 84 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
Thread T7 (MTGS) created by T6 (EE Core) here:
    #0 0xf720601f in pthread_create (/usr/lib/i386-linux-gnu/libasan.so.1+0x1e01f)
    #1 0x897937a in Threading::pxThread::Start() ../common/src/Utilities/ThreadTools.cpp:261
    #2 0x878086f in SysThreadBase::Start() ../pcsx2/System/SysThreadBase.cpp:38
    #3 0x8781773 in SysThreadBase::Resume() ../pcsx2/System/SysThreadBase.cpp:205
    #4 0x81b39a3 in SysCorePlugins::OpenPlugin_GS() ../pcsx2/PluginManager.cpp:1235
    #5 0x846cc4d in AppCorePlugins::OpenPlugin_GS() ../pcsx2/gui/AppCorePlugins.cpp:367
    #6 0x81c1793 in SysCorePlugins::Open(PluginsEnum_t) ../pcsx2/PluginManager.cpp:1311
    #7 0x81b427c in SysCorePlugins::Open() ../pcsx2/PluginManager.cpp:1339
    #8 0x846c57a in AppCorePlugins::Open() ../pcsx2/gui/AppCorePlugins.cpp:355
    #9 0x847902b in AppCoreThread::OnResumeInThread(bool) ../pcsx2/gui/AppCoreThread.cpp:528
    #10 0x8780288 in SysThreadBase::StateCheckInThread() ../pcsx2/System/SysThreadBase.cpp:324
    #11 0x877fe7f in SysCoreThread::StateCheckInThread() ../pcsx2/System/SysCoreThread.cpp:244
    #12 0x877d7e6 in SysCoreThread::ExecuteTaskInThread() ../pcsx2/System/SysCoreThread.cpp:266
    #13 0x897792e in Threading::pxThread::_try_virtual_invoke(void (Threading::pxThread::*)()) ../common/src/Utilities/ThreadTools.cpp:545
    #14 0x897b990 in Threading::pxThread::_internal_execute() ../common/src/Utilities/ThreadTools.cpp:644
    #15 0x897c927 in Threading::pxThread::_internal_callback(void*) ../common/src/Utilities/ThreadTools.cpp:684
    #16 0xf723e51f (/usr/lib/i386-linux-gnu/libasan.so.1+0x5651f)
    #17 0xf59f3d0d in clone (/lib/i386-linux-gnu/i686/cmov/libc.so.6+0xebd0d)

Thread T6 (EE Core) created by T1 (SysExecutor) here:
    #0 0xf720601f in pthread_create (/usr/lib/i386-linux-gnu/libasan.so.1+0x1e01f)
    #1 0x897937a in Threading::pxThread::Start() ../common/src/Utilities/ThreadTools.cpp:261
    #2 0x878086f in SysThreadBase::Start() ../pcsx2/System/SysThreadBase.cpp:38
    #3 0x8781773 in SysThreadBase::Resume() ../pcsx2/System/SysThreadBase.cpp:205
    #4 0x84717ca in AppCoreThread::Resume() ../pcsx2/gui/AppCoreThread.cpp:164
    #5 0x85ed687 in SysExecEvent::_DoInvokeEvent() ../pcsx2/gui/ExecutorThread.cpp:123
    #6 0x85f1c66 in pxEvtQueue::ProcessEvents(std::list<SysExecEvent*, std::allocator<SysExecEvent*> >&, bool) ../pcsx2/gui/ExecutorThread.cpp:218
    #7 0x85f2ddb in pxEvtQueue::ProcessPendingEvents() ../pcsx2/gui/ExecutorThread.cpp:248
    #8 0x85f2ddb in ExecutorThread::ExecuteTaskInThread() ../pcsx2/gui/ExecutorThread.cpp:543
    #9 0x897792e in Threading::pxThread::_try_virtual_invoke(void (Threading::pxThread::*)()) ../common/src/Utilities/ThreadTools.cpp:545
    #10 0x897b990 in Threading::pxThread::_internal_execute() ../common/src/Utilities/ThreadTools.cpp:644
    #11 0x897c927 in Threading::pxThread::_internal_callback(void*) ../common/src/Utilities/ThreadTools.cpp:684
    #12 0xf723e51f (/usr/lib/i386-linux-gnu/libasan.so.1+0x5651f)
    #13 0xf59f3d0d in clone (/lib/i386-linux-gnu/i686/cmov/libc.so.6+0xebd0d)

Thread T1 (SysExecutor) created by T0 here:
    #0 0xf720601f in pthread_create (/usr/lib/i386-linux-gnu/libasan.so.1+0x1e01f)
    #1 0x897937a in Threading::pxThread::Start() ../common/src/Utilities/ThreadTools.cpp:261
    #2 0x84a1f2f in Pcsx2App::OnInit() ../pcsx2/gui/AppInit.cpp:465
    #3 0xf7025c95 in wxEntry(int&, wchar_t**) (/usr/lib/i386-linux-gnu/libwx_baseu-3.0.so.0+0xf4c95)
    #4 0xf7025d32 in wxEntry(int&, char**) (/usr/lib/i386-linux-gnu/libwx_baseu-3.0.so.0+0xf4d32)
    #5 0xf5921a62 in __libc_start_main (/lib/i386-linux-gnu/i686/cmov/libc.so.6+0x19a62)

SUMMARY: AddressSanitizer: stack-buffer-overflow ../common/src/Utilities/ThreadTools.cpp:590 Threading::pxThread::_try_virtual_invoke(void (Threading::pxThread::*)())
Shadow bytes around the buggy address:
  0x3a77fda0: f1 f1 00 04 f4 f4 f2 f2 f2 f2 00 04 f4 f4 f2 f2
  0x3a77fdb0: f2 f2 00 00 04 f4 f2 f2 f2 f2 00 00 00 00 f3 f3
  0x3a77fdc0: f3 f3 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3a77fdd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3a77fde0: 00 00 00 00 00 00 f1 f1 f1 f1 00 04 f4 f4 f3 f3
=>0x3a77fdf0:[f3]f3 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3a77fe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3a77fe10: 00 00 00 00 00 00 00
```

**Comment by turtleli** [comment]
> Edit: I have a strange behavior on debug build. Potentially it is an Nvidia issue or an Asan bug too. I just start in full mode and then shutdown the emulator.

It could be. I get similar ASAN logs on master using no disc. Haven't tested the PR with it yet.

Trimmed ASAN debug log
```
=================================================================
==21713==ERROR: AddressSanitizer: stack-buffer-overflow on address 0xddffee00 at pc 0xf725f402 bp 0xddffed28 sp 0xddffe904
WRITE of size 208 at 0xddffee00 thread T7 (MTGS)
    #0 0xf725f401 in __asan_memcpy (/usr/lib/libasan.so.2+0x97401)
    #1 0xf725f69a in memcpy (/usr/lib/libasan.so.2+0x9769a)
    #2 0xdae02e76  (/usr/lib/libnvidia-glcore.so.370.28+0x14c3e76)
    #3 0xdadfa8e9  (/usr/lib/libnvidia-glcore.so.370.28+0x14bb8e9)
    #4 0xdad29659  (/usr/lib/libnvidia-glcore.so.370.28+0x13ea659)
    #5 0xdad2a892  (/usr/lib/libnvidia-glcore.so.370.28+0x13eb892)
    #6 0xdace7ce9  (/usr/lib/libnvidia-glcore.so.370.28+0x13a8ce9)
    #7 0xdace978a  (/usr/lib/libnvidia-glcore.so.370.28+0x13aa78a)
    #8 0xdb870869  (/usr/lib/libGLX_nvidia.so.0+0x85869)

Address 0xddffee00 is located in stack of thread T7 (MTGS)
SUMMARY: AddressSanitizer: stack-buffer-overflow ??:0 __asan_memcpy
Shadow bytes around the buggy address:
  0x3bbffd70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3bbffd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3bbffd90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3bbffda0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3bbffdb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x3bbffdc0:[f2]f2 00 04 f4 f4 f2 f2 f2 f2 00 00 f4 f4 f2 f2
  0x3bbffdd0: 00 00 00 00 00 00 00 00 00 00 00 00 04 f4 f2 f2
  0x3bbffde0: f2 f2 00 00 00 00 f3 f3 f3 f3 00 00 00 00 00 00
  0x3bbffdf0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3bbffe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x3bbffe10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
Thread T7 (MTGS) created by T6 (EE Core) here:
    #0 0xf71fb1f1 in pthread_create (/usr/lib/libasan.so.2+0x331f1)
    #1 0x87dfadb in Threading::pxThread::Start() /home/turtleli/pcsx2/common/src/Utilities/ThreadTools.cpp:261
    #2 0x8620c1d in SysThreadBase::Start() /home/turtleli/pcsx2/pcsx2/System/SysThreadBase.cpp:38
    #3 0x862199b in SysThreadBase::Resume() /home/turtleli/pcsx2/pcsx2/System/SysThreadBase.cpp:205
    #4 0x812c19b in SysCorePlugins::OpenPlugin_GS() /home/turtleli/pcsx2/pcsx2/PluginManager.cpp:1235
    #5 0x8330540 in AppCorePlugins::OpenPlugin_GS() /home/turtleli/pcsx2/pcsx2/gui/AppCorePlugins.cpp:367
    #6 0x813894d in SysCorePlugins::Open(PluginsEnum_t) /home/turtleli/pcsx2/pcsx2/PluginManager.cpp:1311
    #7 0x812c70e in SysCorePlugins::Open() /home/turtleli/pcsx2/pcsx2/PluginManager.cpp:1339
    #8 0x8330432 in AppCorePlugins::Open() /home/turtleli/pcsx2/pcsx2/gui/AppCorePlugins.cpp:355
    #9 0x861d203 in SysCoreThread::OnResumeInThread(bool) /home/turtleli/pcsx2/pcsx2/System/SysCoreThread.cpp:279
```
ASAN release log (it underflows instead of overflows).
```
=================================================================
==6498==ERROR: AddressSanitizer: stack-buffer-underflow on address 0xd53fed88 at pc 0xf7221402 bp 0xd53fed28 sp 0xd53fe904
WRITE of size 208 at 0xd53fed88 thread T7 (MTGS)
    #0 0xf7221401 in __asan_memcpy (/usr/lib/libasan.so.2+0x97401)
    #1 0xf722169a in memcpy (/usr/lib/libasan.so.2+0x9769a)
    #2 0xd2202e76  (/usr/lib/libnvidia-glcore.so.370.28+0x14c3e76)
    #3 0xd21fa8e9  (/usr/lib/libnvidia-glcore.so.370.28+0x14bb8e9)
    #4 0xd2129659  (/usr/lib/libnvidia-glcore.so.370.28+0x13ea659)
    #5 0xd212a892  (/usr/lib/libnvidia-glcore.so.370.28+0x13eb892)
    #6 0xd20e7ce9  (/usr/lib/libnvidia-glcore.so.370.28+0x13a8ce9)
    #7 0xd20e978a  (/usr/lib/libnvidia-glcore.so.370.28+0x13aa78a)
    #8 0xd2c70869  (/usr/lib/libGLX_nvidia.so.0+0x85869)

Address 0xd53fed88 is located in stack of thread T7 (MTGS)==6498==AddressSanitizer CHECK failed: ../../../../gcc-5.4.0/libsanitizer/asan/asan_thread.cc:231 "((ptr[0] == kCurrentStackFrameMagic)) != (0)" (0x0, 0x0)
    #0 0xf723805d  (/usr/lib/libasan.so.2+0xae05d)
    #1 0xf723d9cc in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/usr/lib/libasan.so.2+0xb39cc)
    #2 0xf723b461  (/usr/lib/libasan.so.2+0xb1461)
    #3 0xf7234798  (/usr/lib/libasan.so.2+0xaa798)
    #4 0xf7234e2b  (/usr/lib/libasan.so.2+0xaae2b)
    #5 0xf7236c18 in __asan_report_error (/usr/lib/libasan.so.2+0xacc18)
    #6 0xf722142e in __asan_memcpy (/usr/lib/libasan.so.2+0x9742e)
    #7 0xf722169a in memcpy (/usr/lib/libasan.so.2+0x9769a)
    #8 0xd2202e76  (/usr/lib/libnvidia-glcore.so.370.28+0x14c3e76)
```

**Comment by gregory38** [comment]
I'm fine with the code I managed to boot 2 games. My reader does some bad noises (but I never use it so my reader could suck too). Feel free to merge it, if you're ready.

**Pull request merged**